### PR TITLE
FGNHG CLI changes

### PR DESCRIPTION
**Pull request by kktheballer**
1. Improve CLI output format
2. Adapt CLI to match-mode based next-hop alias mapping

#### What I did
- Improved CLI output format
- Enabled NH alias mapping via FG_NHG entry in config_db

#### How to verify it
Test on DUT

#### Previous command output (if the output of a command-line utility has changed)
Please refer to tests/fgnhg_test.py

#### New command output (if the output of a command-line utility has changed)
Please refer to tests/fgnhg_test.py

### CODE REVIEW

**Comment by liat-grozovik** [comment]
@kktheballer could you please fix the LGTM alerts? beside that it looks good

**Review by liat-grozovik — DISMISSED** [review]

File: tests/fgnhg_test.py [hunk 1]
```diff
@@ -9,187 +9,45 @@
...
-| fc:5::/128      | 200:200:200:200::4 |
-|                 | 200:200:200:200::5 |
-+-----------------+--------------------+
-| 100.50.25.12/32 | 200.200.200.4      |
-|                 | 200.200.200.5      |
-+-----------------+--------------------+
+FG NHG Prefix    Active Next Hops
+---------------  --------------------------------------
+fc:5::/128       200:200:200:200::4, 200:200:200:200::5
+100.50.25.12/32  200.200.200.4, 200.200.200.5
```
Comment (anish-n) [inline tests/fgnhg_test.py:38]: This may end up as a lot of output on a single line, maybe we can limit how many we print on one line or separate each next-hop with a new line(as it was originally).

File: tests/fgnhg_test.py [hunk 2]
```diff
@@ -9,187 +9,45 @@
...
-|                 |                    | 24             |
-|                 |                    | 25             |
-|                 |                    | 26             |
-|                 |                    | 27             |
-|                 |                    | 28             |
-|                 |                    | 29             |
-|                 |                    | 30             |
-|                 |                    | 31             |
-+-----------------+--------------------+----------------+
+FG NHG Prefix    Next Hop            Hash buckets
```
Comment (anish-n) [inline tests/fgnhg_test.py:12]: We may end up with a lot of hash buckets on one line, making it hard to parse, perhaps we can limit how many hash buckets we print on 1 line, eg at most 30-50chars on one line

File: tests/fgnhg_test.py [hunk 3]
```diff
@@ -9,187 +9,45 @@
...
-|                 |               | 25             |
-|                 |               | 26             |
-|                 |               | 27             |
-|                 |               | 28             |
-|                 |               | 29             |
-|                 |               | 30             |
-|                 |               | 31             |
-+-----------------+---------------+----------------+
+Alias     Next Hop       Hash buckets
+--------  -------------  ----------------------------
```
Comment (anish-n) [inline tests/fgnhg_test.py:22]: Why are there 2 differing outputs for show fgnhg hash_view, one which contains an alias column and one which contains a prefix column, could we have both columns displayed perhaps?

File: show/fgnhg.py [hunk 4]
```diff
@@ -32,12 +25,22 @@ def active_hops(nhg):
...
     t_dict = {}
-    header = ["FG_NHG_PREFIX", "Active Next Hops"]
+    header = ["FG NHG Prefix", "Active Next Hops"]
     table = []
     output_dict = {}
 
+    try:
+        table_keys = sorted(state_db.keys(state_db.STATE_DB, _hash))
+    except BaseException as e:
+        click.echo("FG_ROUTE_TABLE does not exist!")
```
Comment (nazariig) [inline show/fgnhg.py:36]: @kktheballer usually in `Click` we use the next patterns:
```python
def cli_show_help(ctx):
    click.echo(ctx.get_help())
    ctx.exit(EXIT_SUCCESS)

def cli_fail(ctx, msg):
    ctx.fail(msg)

def cli_abort(ctx, msg):
    click.echo("Error: " + msg + ". Aborting...")
    ctx.abort()

def cli_exit(ctx, msg):
    click.echo("Info: {}.".format(msg))
    ctx.exit(EXIT_SUCCESS)
```

File: show/fgnhg.py [hunk 5]
```diff
@@ -52,47 +55,56 @@ def active_hops(nhg):
...
+        alias_list = []
+        nexthop_alias = {}
+        output_list = []
+
+        for nexthop, nexthop_metadata in fg_nhg_member_table.items():
+            alias_list.append(nexthop_metadata['FG_NHG'])
+            nexthop_alias[nexthop] = nexthop_metadata['FG_NHG']
+
+        if nhg not in alias_list:
+            click.echo ("Please provide a valid NHG alias")
```
Comment (nazariig) [inline show/fgnhg.py:78]: @kktheballer please remove the extra space

**Comment by nazariig** [comment]
@kktheballer can you please mark the fixed comments as resolved?

File: show/fgnhg.py [hunk 6]
```diff
@@ -52,61 +52,78 @@ def active_hops(nhg):
...
+                ctx.fail ("FG_ROUTE table likely does not contain the required entries")
+                sys.exit(1)
 
-                nhg_prefix_report = (nhg_prefix.split("|")[1])
-                formatted_nhps = ','.replace(',', '\n').join(output_dict[nhg_prefix])
-                table.append([nhg_prefix_report, formatted_nhps])
-                click.echo(tabulate(table, header, tablefmt="grid"))
+            nhg_prefix_report = nhip_prefix_map[output_list[0]].split("|")[1]
+            formatted_output_list = ','.replace(',', '\n').join(output_list)
+            table.append([nhg_prefix_report, formatted_output_list])
```
Comment (nazariig) [inline show/fgnhg.py:95]: @kktheballer can you please also align `tablefmt` for all FG CLI to make it SONiC common?

**Example:**
```
root@sonic:/home/admin# show interfaces status
      Interface            Lanes    Speed    MTU    FEC    Alias             Vlan    Oper    Admin    Type    Asym PFC
---------------  ---------------  -------  -----  -----  -------  ---------------  ------  -------  ------  ----------
      Ethernet0          0,1,2,3     100G   9100    N/A     etp1           routed      up       up     N/A         N/A
      Ethernet4          4,5,6,7     100G   9100    N/A     etp2           routed      up       up     N/A         N/A
      Ethernet8        8,9,10,11     100G   9100    N/A     etp3           routed      up       up     N/A         N/A
     Ethernet12      12,13,14,15     100G   9100    N/A     etp4           routed      up       up     N/A         N/A
     Ethernet16      16,17,18,19     100G   9100    N/A     etp5           routed      up       up     N/A         N/A
     Ethernet20      20,21,22,23     100G   9100    N/A     etp6           routed      up       up     N/A         N/A
     Ethernet24      24,25,26,27     100G   9100    N/A     etp7           routed      up       up     N/A         N/A
     Ethernet28      28,29,30,31     100G   9100    N/A     etp8           routed      up       up     N/A         N/A
     Ethernet32      32,33,34,35     100G   9100    N/A     etp9           routed      up       up     N/A         N/A
     Ethernet36      36,37,38,39     100G   9100    N/A    etp10           routed      up       up     N/A         N/A
     Ethernet40      40,41,42,43     100G   9100    N/A    etp11           routed      up       up     N/A         N/A
     Ethernet44      44,45,46,47     100G   9100    N/A    etp12           routed      up       up     N/A         N/A
     Ethernet48      48,49,50,51     100G   9100    N/A    etp13           routed      up       up     N/A         N/A
     Ethernet52      52,53,54,55     100G   9100    N/A    etp14           routed      up       up     N/A         N/A
     Ethernet56      56,57,58,59     100G   9100    N/A    etp15           routed      up       up     N/A         N/A
     Ethernet60      60,61,62,63     100G   9100    N/A    etp16           routed      up       up     N/A         N/A
     Ethernet64      64,65,66,67     100G   9100    N/A    etp17           routed      up       up     N/A         N/A
     Ethernet68      68,69,70,71     100G   9100    N/A    etp18           routed      up       up     N/A         N/A
     Ethernet72      72,73,74,75     100G   9100    N/A    etp19           routed      up       up     N/A         N/A
     Ethernet76      76,77,78,79     100G   9100    N/A    etp20           routed      up       up     N/A         N/A
     Ethernet80      80,81,82,83     100G   9100    N/A    etp21           routed      up       up     N/A         N/A
     Ethernet84      84,85,86,87     100G   9100    N/A    etp22           routed      up       up     N/A         N/A
     Ethernet88      88,89,90,91     100G   9100    N/A    etp23           routed      up       up     N/A         N/A
     Ethernet92      92,93,94,95     100G   9100    N/A    etp24           routed      up       up     N/A         N/A
     Ethernet96      96,97,98,99     100G   9100    N/A    etp25           routed      up       up     N/A         N/A
    Ethernet100  100,101,102,103     100G   9100    N/A    etp26           routed      up       up     N/A         N/A
    Ethernet104  104,105,106,107     100G   9100    N/A    etp27           routed      up       up     N/A         N/A
    Ethernet108  108,109,110,111     100G   9100    N/A    etp28           routed      up       up     N/A         N/A
    Ethernet112  112,113,114,115     100G   9100    N/A    etp29           routed      up       up     N/A         N/A
    Ethernet116  116,117,118,119     100G   9100    N/A    etp30           routed      up       up     N/A         N/A
    Ethernet120  120,121,122,123     100G   9100    N/A    etp31  PortChannel0001      up       up     N/A         N/A
    Ethernet124  124,125,126,127     100G   9100    N/A    etp32  PortChannel0002      up       up     N/A         N/A
PortChannel0001              N/A     100G   9100    N/A      N/A           routed    down     down     N/A         N/A
PortChannel0002              N/A     100G   9100    N/A      N/A           routed    down     down     N/A         N/A
```

**Comment by nazariig** [comment]
@kktheballer please rebase these changes on top of latest `master` and re-run checks

**Comment by nazariig** [comment]
@kktheballer can you please rebase to the latest `master`?

File: show/fgnhg.py [hunk 7]
```diff
@@ -12,31 +11,32 @@ def fgnhg():
...
+    header = ["FG NHG Prefix", "Active Next Hops"]
     table = []
     output_dict = {}
+    ctx = click.get_current_context()
+
+    try:
+        table_keys = sorted(state_db.keys(state_db.STATE_DB, _hash))
+    except Exception as e:
+        ctx.fail("FG_ROUTE_TABLE does not exist!")
+        exit()
```
Comment (nazariig) [inline show/fgnhg.py:35]: @kktheballer seems to be unreachable section

File: show/fgnhg.py [hunk 8]
```diff
@@ -12,31 +11,32 @@ def fgnhg():
...
+
+    if table_keys is None:
+        ctx.fail("FG_ROUTE_TABLE does not exist!")
+        exit()
```
Comment (nazariig) [inline show/fgnhg.py:39]: @kktheballer seems to be unreachable section

File: show/fgnhg.py [hunk 9]
```diff
@@ -52,61 +52,78 @@ def active_hops(nhg):
...
-                    nhg_prefix = "FG_ROUTE_TABLE|" + nhg_prefix
-                    t_dict = state_db.get_all(state_db.STATE_DB, nhg_prefix)
-
+        nhip_prefix_map = {}
+        header = ["FG NHG Prefix", "Active Next Hops"]
+        try:
+            fg_nhg_member_table = config_db.get_table('FG_NHG_MEMBER')
+        except Exception as e: 
+            ctx.fail("FG_NHG_MEMBER entries not present in config_db")
+            exit()
```
Comment (nazariig) [inline show/fgnhg.py:64]: @kktheballer seems to be unreachable section

File: show/fgnhg.py [hunk 10]
```diff
@@ -52,61 +52,78 @@ def active_hops(nhg):
...
     table = []
     output_dict = {}
     bank_dict = {}
+    ctx = click.get_current_context()
+
+    try:
+        table_keys = sorted(state_db.keys(state_db.STATE_DB, _hash))
+    except Exception as e: 
+        ctx.fail("FG_ROUTE_TABLE does not exist!")
+        click.echo(e)
```
Comment (nazariig) [inline show/fgnhg.py:121]: @kktheballer seems to be unreachable section

File: show/fgnhg.py [hunk 11]
```diff
@@ -52,61 +52,78 @@ def active_hops(nhg):
...
+        exit()
+
+    if table_keys is None:
+        ctx.fail("FG_ROUTE_TABLE does not exist!")
+        exit()
```
Comment (nazariig) [inline show/fgnhg.py:126]: @kktheballer seems to be unreachable section

File: show/fgnhg.py [hunk 12]
```diff
@@ -129,41 +146,94 @@ def hash_view(nhg):
...
-                nhg_prefix_report = (nhg_prefix.split("|")[1])
                 bank_dict = OrderedDict(sorted(bank_dict.items()))
+                output_bank_dict = {}
+                for nexthop, banks  in bank_dict.items():
+                    if nexthop in nexthop_alias:
+                        if nexthop_alias[nexthop] == nhg:
+                            output_bank_dict[nexthop] = banks
+                    else:
+                        ctx.fail ("state_db and config_db have FGNHG prefix config mismatch. Check device config!");
+                        sys.exit(1)
```
Comment (nazariig) [inline show/fgnhg.py:214]: @kktheballer seems to be unreachable section

File: show/fgnhg.py [hunk 13]
```diff
@@ -129,41 +146,94 @@ def hash_view(nhg):
...
 
-                nhg_prefix_report = (nhg_prefix.split("|")[1])
                 bank_dict = OrderedDict(sorted(bank_dict.items()))
+                output_bank_dict = {}
+                for nexthop, banks  in bank_dict.items():
+                    if nexthop in nexthop_alias:
+                        if nexthop_alias[nexthop] == nhg:
+                            output_bank_dict[nexthop] = banks
+                    else:
+                        ctx.fail ("state_db and config_db have FGNHG prefix config mismatch. Check device config!");
```
Comment (nazariig) [inline show/fgnhg.py:213]: @kktheballer please remove redundant space

File: show/fgnhg.py [hunk 14]
```diff
@@ -52,61 +52,78 @@ def active_hops(nhg):
...
 
-                nhg_prefix_report = (nhg_prefix.split("|")[1])
-                formatted_nhps = ','.replace(',', '\n').join(output_dict[nhg_prefix])
-                table.append([nhg_prefix_report, formatted_nhps])
-                click.echo(tabulate(table, header, tablefmt="grid"))
+            nhg_prefix_report = nhip_prefix_map[output_list[0]].split("|")[1]
+            formatted_output_list = ','.replace(',', '\n').join(output_list)
+            table.append([nhg_prefix_report, formatted_output_list])
 
+            click.echo(tabulate(table, header, tablefmt = "simple"))
```
Comment (nazariig) [inline show/fgnhg.py:97]:
```suggestion
            click.echo(tabulate(table, header, tablefmt="simple"))
```

File: show/fgnhg.py [hunk 15]
```diff
@@ -129,41 +146,94 @@ def hash_view(nhg):
...
+                        displayed_banks.append(str(bankid))
 
-                vals = sorted(set([val for val in t_dict.values()]))
+                for i in range (0, len(displayed_banks), 8):
+                    bank_output = bank_output + " ".join(displayed_banks[i:i+8]) + "\n" 
 
+                bank_output = bank_output + "\n"
+                table.append([nhg_prefix_report, nhip, bank_output])
+
+        click.echo(tabulate(table, header, tablefmt = "simple"))
```
Comment (nazariig) [inline show/fgnhg.py:168]:
```suggestion
        click.echo(tabulate(table, header, tablefmt="simple"))
```

File: show/fgnhg.py [hunk 16]
```diff
@@ -129,41 +146,94 @@ def hash_view(nhg):
...
+                    for i in range (0, len(displayed_banks), 8):
+                        bank_output = bank_output + " ".join(displayed_banks[i:i+8]) + "\n" 
 
-                for nhip, val in bank_dict.items():
-                    formatted_banks = ','.replace(',', '\n').join(bank_dict[nhip])
-                    table.append([nhg_prefix_report, nhip, formatted_banks])
+                    table.append([nhg_prefix_report, nhip, bank_output])
 
-                click.echo(tabulate(table, header, tablefmt="grid"))
+            click.echo(tabulate(table, header, tablefmt = "simple"))
```
Comment (nazariig) [inline show/fgnhg.py:239]:
```suggestion
            click.echo(tabulate(table, header, tablefmt="simple"))
```

File: show/fgnhg.py (hunk 12 above)
Comment (nazariig) [reply show/fgnhg.py:214]: @kktheballer `sys.exit(1)` is not reachable, since `ctx.fail` will exit context immediately

File: show/fgnhg.py [hunk 17]
```diff
@@ -52,61 +50,73 @@ def active_hops(nhg):
...
+            ctx.fail("FG_NHG_MEMBER entries not present in config_db")
+        alias_list = []
+        nexthop_alias = {}
+        output_list = []
+
+        for nexthop, nexthop_metadata in fg_nhg_member_table.items():
+            alias_list.append(nexthop_metadata['FG_NHG'])
+            nexthop_alias[nexthop] = nexthop_metadata['FG_NHG']
+        if nhg not in alias_list:
+            ctx.fail ("Please provide a valid NHG alias")
```
Comment (nazariig) [inline show/fgnhg.py:70]: @kktheballer please remove redundant space

File: show/fgnhg.py [hunk 18]
```diff
@@ -52,61 +50,73 @@ def active_hops(nhg):
...
-                    if nhg_prefix in output_dict:
-                        output_dict[nhg_prefix].append(nh_ip.split("@")[0])
+                    nhip_prefix_map[nh_ip.split("@")[0]] = nhg_prefix
+                   
+                    if nh_ip.split("@")[0] in nexthop_alias:
+                        if nexthop_alias[nh_ip.split("@")[0]] == nhg:
+                            output_list.append(nh_ip.split("@")[0])
                     else:
-                        output_dict[nhg_prefix] = [nh_ip.split("@")[0]]
+                        ctx.fail ("state_db and config_db have FGNHG prefix config mismatch. Check device config!");
```
Comment (nazariig) [inline show/fgnhg.py:83]: @kktheballer please remove redundant space

File: show/fgnhg.py [hunk 19]
```diff
@@ -52,61 +50,73 @@ def active_hops(nhg):
...
+                output_list = sorted(output_list)
+            if not output_list:
+                ctx.fail ("FG_ROUTE table likely does not contain the required entries")
```
Comment (nazariig) [inline show/fgnhg.py:86]: @kktheballer please remove redundant space

File: show/fgnhg.py [hunk 20]
```diff
@@ -129,41 +139,92 @@ def hash_view(nhg):
...
+
+        alias_list = []
+        nexthop_alias = {}
+
+        for nexthop, nexthop_metadata in fg_nhg_member_table.items():
+            alias_list.append(nexthop_metadata['FG_NHG'])
+            nexthop_alias[nexthop] = nexthop_metadata['FG_NHG']
+
+        if nhg not in alias_list:
+            ctx.fail ("Please provide a valid NHG alias")
```
Comment (nazariig) [inline show/fgnhg.py:179]: @kktheballer please remove redundant space

File: show/fgnhg.py [hunk 21]
```diff
@@ -1,6 +1,5 @@
-import ipaddress
 from collections import OrderedDict
-
+import sys
```
Comment (nazariig) [inline show/fgnhg.py:2]: @kktheballer Import of `sys` is not used.

**Review by nazariig — CHANGES_REQUESTED** [review]

File: show/fgnhg.py (hunk 17 above)
Comment (nazariig) [reply show/fgnhg.py:63]: @kktheballer this hasn't been fixed

File: show/fgnhg.py (hunk 18 above)
Comment (nazariig) [reply show/fgnhg.py:75]: @kktheballer this hasn't been fixed

File: show/fgnhg.py (hunk 19 above)
Comment (nazariig) [reply show/fgnhg.py:78]: @kktheballer this hasn't been fixed

File: show/fgnhg.py (hunk 10 above)
Comment (nazariig) [reply show/fgnhg.py:105]: @kktheballer this hasn't been fixed

File: show/fgnhg.py [hunk 22]
```diff
@@ -129,41 +146,94 @@ def hash_view(nhg):
...
+
+        click.echo(tabulate(table, header, tablefmt = "simple"))
+
+    else:
+        header = ["FG NHG Prefix", "Next Hop", "Hash buckets"]
+        try:
+            fg_nhg_member_table = config_db.get_table('FG_NHG_MEMBER')
+        except Exception as e: 
+            ctx.fail("FG_NHG_MEMBER entries not present in config_db")
+            click.echo(e)
```
[the remark this answers is not in the document]
Comment (nazariig) [reply show/fgnhg.py:147]: @kktheballer this hasn't been fixed

File: show/fgnhg.py (hunk 20 above)
Comment (nazariig) [reply show/fgnhg.py:154]: @kktheballer this hasn't been fixed

File: show/fgnhg.py [hunk 23]
```diff
@@ -47,123 +41,156 @@ def active_hops(nhg):
...
-                nhg_prefix_report = (nhg_prefix.split("|")[1])
                 bank_dict = OrderedDict(sorted(bank_dict.items()))
-
-                for nhip, val in bank_dict.items():
-                    formatted_banks = ','.replace(',', '\n').join(bank_dict[nhip])
-                    table.append([nhg_prefix_report, nhip, formatted_banks])
-
-                click.echo(tabulate(table, header, tablefmt="grid"))
+                output_bank_dict = {}
+                for nexthop, banks  in bank_dict.items():
```
Comment (nazariig) [inline show/fgnhg.py:172]:
```suggestion
                for nexthop, banks in bank_dict.items():
```

File: show/fgnhg.py [hunk 24]
```diff
@@ -47,123 +41,154 @@ def active_hops(nhg):
...
+        except Exception as e: 
+            ctx.fail("FG_NHG_MEMBER entries not present in config_db")
+        alias_list = []
+        nexthop_alias = {}
+        output_list = []
+        for nexthop, nexthop_metadata in fg_nhg_member_table.items():
+            alias_list.append(nexthop_metadata['FG_NHG'])
+            nexthop_alias[nexthop] = nexthop_metadata['FG_NHG']
+        if nhg not in alias_list:
+            ctx.fail ("Please provide a valid NHG alias")
```
Comment (nazariig) [inline show/fgnhg.py:63]:
```suggestion
            ctx.fail("Please provide a valid NHG alias")
```

File: show/fgnhg.py [hunk 25]
```diff
@@ -47,123 +41,154 @@ def active_hops(nhg):
...
+                            output_list.append(nh_ip.split("@")[0])
                     else:
-                        output_dict[nhg_prefix] = [nh_ip.split("@")[0]]
-
-                nhg_prefix_report = (nhg_prefix.split("|")[1])
-                formatted_nhps = ','.replace(',', '\n').join(output_dict[nhg_prefix])
-                table.append([nhg_prefix_report, formatted_nhps])
-                click.echo(tabulate(table, header, tablefmt="grid"))
-
+                        ctx.fail ("state_db and config_db have FGNHG prefix config mismatch. Check device config!");
```
Comment (nazariig) [inline show/fgnhg.py:75]:
```suggestion
                        ctx.fail("state_db and config_db have FGNHG prefix config mismatch. Check device config!");
```

File: show/fgnhg.py [hunk 26]
```diff
@@ -47,123 +41,154 @@ def active_hops(nhg):
...
+                output_list = sorted(output_list)
+            if not output_list:
+                ctx.fail ("FG_ROUTE table likely does not contain the required entries")
```
Comment (nazariig) [inline show/fgnhg.py:78]:
```suggestion
                ctx.fail("FG_ROUTE table likely does not contain the required entries")
```

File: show/fgnhg.py [hunk 27]
```diff
@@ -47,123 +41,154 @@ def active_hops(nhg):
...
+            fg_nhg_member_table = config_db.get_table('FG_NHG_MEMBER')
+        except Exception as e: 
+            ctx.fail("FG_NHG_MEMBER entries not present in config_db")
+        alias_list = []
+        nexthop_alias = {}
+        for nexthop, nexthop_metadata in fg_nhg_member_table.items():
+            alias_list.append(nexthop_metadata['FG_NHG'])
+            nexthop_alias[nexthop] = nexthop_metadata['FG_NHG']
+        if nhg not in alias_list:
+            ctx.fail ("Please provide a valid NHG alias")
```
Comment (nazariig) [inline show/fgnhg.py:152]:
```suggestion
            ctx.fail("Please provide a valid NHG alias")
```

File: show/fgnhg.py [hunk 28]
```diff
@@ -47,123 +41,154 @@ def active_hops(nhg):
...
-                    table.append([nhg_prefix_report, nhip, formatted_banks])
-
-                click.echo(tabulate(table, header, tablefmt="grid"))
+                output_bank_dict = {}
+                for nexthop, banks in bank_dict.items():
+                    if nexthop in nexthop_alias:
+                        if nexthop_alias[nexthop] == nhg:
+                            output_bank_dict[nexthop] = banks
+                    else:
+                        ctx.fail ("state_db and config_db have FGNHG prefix config mismatch. Check device config!");
```
Comment (nazariig) [inline show/fgnhg.py:175]:
```suggestion
                        ctx.fail("state_db and config_db have FGNHG prefix config mismatch. Check device config!");
```

**Review by nazariig — DISMISSED** [review]

File: show/fgnhg.py [hunk 29]
```diff
@@ -47,123 +41,154 @@ def active_hops(nhg):
...
-                nhg_prefix_report = (nhg_prefix.split("|")[1])
-                formatted_nhps = ','.replace(',', '\n').join(output_dict[nhg_prefix])
-                table.append([nhg_prefix_report, formatted_nhps])
-                click.echo(tabulate(table, header, tablefmt="grid"))
-
+                        ctx.fail("state_db and config_db have FGNHG prefix config mismatch. Check device config!");
+                output_list = sorted(output_list)
+            if not output_list:
+                ctx.fail("FG_ROUTE table likely does not contain the required entries")
+            nhg_prefix_report = nhip_prefix_map[output_list[0]].split("|")[1]
```
Comment (anish-n) [inline show/fgnhg.py:79]: nit: can we set nhg_prefix = nhg_prefix.split("|")[1] in the initial parsing itself(line 65-66)? I don't think we are using nhg_prefix.split("|")[0] anywhere in any case.

File: show/fgnhg.py [hunk 30]
```diff
@@ -47,123 +41,154 @@ def active_hops(nhg):
...
+            formatted_output_list = ','.replace(',', '\n').join(output_list)
+            table.append([nhg_prefix_report, formatted_output_list])
```
Comment (anish-n) [inline show/fgnhg.py:81]: Is the output format updated in the HLD? https://github.com/Azure/SONiC/blob/master/doc/ecmp/fine_grained_next_hop_hld.md#232-cli-show-fgnhg-active-hops-fg-nhg-group-name-output-format

File: tests/fgnhg_test.py [hunk 31]
```diff
@@ -9,187 +9,49 @@
...
-|                 |                    | 26             |
-|                 |                    | 27             |
-|                 |                    | 28             |
-|                 |                    | 29             |
-|                 |                    | 30             |
-|                 |                    | 31             |
-+-----------------+--------------------+----------------+
+FG NHG Prefix    Next Hop            Hash buckets
+---------------  ------------------  ------------------------------
+100.50.25.12/32  200.200.200.4       0   1   2   3   4   5   6   7
```
Comment (anish-n) [inline tests/fgnhg_test.py:14]: Could we add one unit test where there are multiple prefixes utilizing the same next-hop IP to check if the output looks good in that scenario?

example:
  100.50.25.12/32 nh 200.200.200.5
   100.50.25.13/32 nh 200.200.200.5

File: show/fgnhg.py [hunk 32]
```diff
@@ -47,123 +41,154 @@ def active_hops(nhg):
...
-                formatted_banks = ','.replace(',', '\n').join(bank_dict[nhip])
-                table.append([nhg_prefix_report, nhip, formatted_banks])
-
-        click.echo(tabulate(table, header, tablefmt="grid"))
-
+                displayed_banks = []
+                bank_output = ""
+                formatted_banks = (bank_dict[nhip])
+                for bankid in formatted_banks:
+                    if (len(str(bankid)) == 1):
```
Comment (anish-n) [inline show/fgnhg.py:127]: Can we add a comment here to explain this logic, I understand that we are doing this to keep the width and printing of bucket # clean with adjusted spacing, but maybe good to have a small comment on it here.

File: show/fgnhg.py [hunk 33]
```diff
@@ -47,123 +41,154 @@ def active_hops(nhg):
...
-                        mod_key = key.split("|")[1].split("/")[0]
-                        mod_nhg_prefix = nhg_prefix.split("/")[0]
-                        if ipaddress.ip_address(mod_key).exploded == ipaddress.ip_address(mod_nhg_prefix).exploded:
-                            t_dict = state_db.get_all(state_db.STATE_DB, key)
-                    nhg_prefix = "FG_ROUTE_TABLE|" + nhg_prefix
-                else:
-                    nhg_prefix = "FG_ROUTE_TABLE|" + nhg_prefix
-                    t_dict = state_db.get_all(state_db.STATE_DB, nhg_prefix)
-
+        header = ["FG NHG Prefix", "Next Hop", "Hash buckets"]
```
Comment (anish-n) [inline show/fgnhg.py:141]: Do we need an overridden header, seems to be the same set of fields

**Review by anish-n — DISMISSED** [review]
Approved with some nit/suggested improvements.

**Review by nazariig — APPROVED** [review]

**Review by anish-n — APPROVED** [review]

File: show/fgnhg.py (hunk 30 above)
Comment (kktheballer) [reply show/fgnhg.py:81]: Yes there is a PR for that also. https://github.com/Azure/SONiC/pull/854

File: show/fgnhg.py [hunk 34]
```diff
@@ -47,123 +41,154 @@ def active_hops(nhg):
...
-                        mod_nhg_prefix = nhg_prefix.split("/")[0]
-                        if ipaddress.ip_address(mod_key).exploded == ipaddress.ip_address(mod_nhg_prefix).exploded:
-                            t_dict = state_db.get_all(state_db.STATE_DB, key)
-                    nhg_prefix = "FG_ROUTE_TABLE|" + nhg_prefix
-                else:
-                    nhg_prefix = "FG_ROUTE_TABLE|" + nhg_prefix
-                    t_dict = state_db.get_all(state_db.STATE_DB, nhg_prefix)
-
+        nhip_prefix_map = {}
+        header = ["FG NHG Prefix", "Active Next Hops"]
```
[the remark this answers is not in the document]
Comment (kktheballer) [reply show/fgnhg.py:51]: got it

File: show/fgnhg.py (hunk 29 above)
Comment (kktheballer) [reply show/fgnhg.py:79]: sure can do that

File: show/fgnhg.py (hunk 33 above)
Comment (kktheballer) [reply show/fgnhg.py:141]: got it can address the redundant header decls

File: show/fgnhg.py (hunk 32 above)
Comment (kktheballer) [reply show/fgnhg.py:127]: sure, will do

File: tests/fgnhg_test.py (hunk 31 above)
Comment (kktheballer) [reply tests/fgnhg_test.py:14]: will add this UT in a follow up PR

**Review by liat-grozovik — APPROVED** [review]

**Comment by liat-grozovik** [comment]
@kktheballer could you please update the command reference doc as well with the recent changes?
